File: packages/logger/src/telemetry-manager.ts
import { createConsoleSink } from './console-sink.js'
import { createNoopTelemetry } from './noop.js'
import { resolvePreset, type TelemetryPreset } from './presets.js'
import { redact } from './redact.js'
import {
  LEVEL_ORDER,
  type LogContext,
  type LogLevel,
  type LogRecord,
  type Span,
  type SpanRecord,
  type Telemetry,
  type TelemetryConfig,
  type TelemetrySink,
} from './types.js'

/**
 * createTelemetry — creates a telemetry manager that fans records out to
 * registered sinks. Manager/provider pattern, mirroring `createAI`.
 *
 * - `enabled: false` -> tree-shakeable noop (zero emissions, no engines).
 * - no `endpoint` -> console-only transport.
 * - `endpoint` set -> async Faro engine is registered when the optional
 *   peers exist; console stays as a safe fallback until then.
 *
 * The returned object IS a logger (root context = `{}`); `child()` derives
 * loggers with bound context (sessionId / interviewId / turnId / ...).
 */
export function createTelemetry(config: TelemetryConfig): Telemetry {
  if (config.enabled === false) {
    return createNoopTelemetry()
  }

  const preset = resolvePreset(config.env)
  const sampleRate = config.sampleRate ?? preset.sampleRate
  const redactKeys = config.redactKeys ?? []

  const sinks = new Map<string, TelemetrySink>()
  /** Insertion-ordered sink names. */
  const sinkOrder: string[] = []
  /** Batch buffer (production preset only). */
  const buffer: Array<{ kind: 'log'; record: LogRecord } | { kind: 'span'; record: SpanRecord }> = []

  function addSinkInternal(sink: TelemetrySink): void {
    if (sinks.has(sink.name)) {
      sinks.set(sink.name, sink)
    } else {
      sinks.set(sink.name, sink)
      sinkOrder.push(sink.name)
    }
  }

  // Console transport is always present as the zero-dependency baseline.
  addSinkInternal(createConsoleSink({ pretty: preset.pretty }))

  // When an endpoint is configured, attempt to attach the Faro engine. The
  // import is lazy + dynamic so the optional peers never become required and
  // Faro names stay out of the public API.
  if (config.endpoint) {
    const endpoint = config.endpoint
    void import('./faro-engine.js')
      .then(({ createFaroSink }) => createFaroSink({ app: config.app, endpoint }))
      .then((faro) => {
        if (faro) addSinkInternal(faro)
      })
      .catch(() => {
        /* peers absent or init failed — console remains */
      })
  }

  function shouldSample(): boolean {
    if (sampleRate >= 1) return true
    if (sampleRate <= 0) return false
    return Math.random() < sampleRate
  }

  function dispatch(
    entry: { kind: 'log'; record: LogRecord } | { kind: 'span'; record: SpanRecord },
  ): void {
    if (preset.batch) {
      buffer.push(entry)
      if (buffer.length >= preset.batchSize) {
        void flushBuffer()
      }
      return
    }
    deliver(entry)
  }

  function deliver(
    entry: { kind: 'log'; record: LogRecord } | { kind: 'span'; record: SpanRecord },
  ): void {
    for (const name of sinkOrder) {
      const sink = sinks.get(name)
      if (!sink) continue
      if (entry.kind === 'log') sink.log(entry.record)
      else sink.span(entry.record)
    }
  }

  async function flushBuffer(): Promise<void> {
    if (buffer.length > 0) {
      const pending = buffer.splice(0, buffer.length)
      for (const entry of pending) deliver(entry)
    }
    await Promise.all(
      sinkOrder.map((name) => sinks.get(name)?.flush() ?? Promise.resolve()),
    )
  }

  // ---- page-exit beacon flush (production preset) -------------------------
  // Browser only — `navigator.sendBeacon` is used by the underlying engine
  // transport; here we just trigger a flush on the page-exit signals.
  const root = globalThis as {
    addEventListener?: (type: string, listener: () => void) => void
    document?: { visibilityState?: string }
    navigator?: { sendBeacon?: unknown }
  }
  if (preset.beaconFlush && typeof root.addEventListener === 'function') {
    const onExit = (): void => {
      void flushBuffer()
    }
    root.addEventListener('pagehide', onExit)
    root.addEventListener('visibilitychange', () => {
      if (root.document?.visibilityState === 'hidden') onExit()
    })
  }

  function makeLogger(boundContext: LogContext): Telemetry {
    function emit(level: LogLevel, message: string, context?: LogContext): void {
      if (LEVEL_ORDER[level] < LEVEL_ORDER[preset.minLevel]) return
      if (!shouldSample()) return
      const merged = redact({ ...boundContext, ...context }, redactKeys)
      const record: LogRecord = {
        level,
        message,
        timestamp: Date.now(),
        app: config.app,
        env: config.env,
        context: merged,
      }
      dispatch({ kind: 'log', record })
    }

    const logger: Telemetry = {
      debug(message, context?) {
        emit('debug', message, context)
      },
      info(message, context?) {
        emit('info', message, context)
      },
      warn(message, context?) {
        emit('warn', message, context)
      },
      error(message, context?) {
        emit('error', message, context)
      },
      fatal(message, context?) {
        emit('fatal', message, context)
      },

      child(context: LogContext): Telemetry {
        return makeLogger({ ...boundContext, ...context })
      },

      startSpan(name: string, attributes?: LogContext): Span {
        const startTime = Date.now()
        let ended = false
        return {
          end(opts?: { error?: unknown; attributes?: LogContext }): void {
            if (ended) return
            ended = true
            const endTime = Date.now()
            const merged = redact(
              { ...boundContext, ...attributes, ...opts?.attributes },
              redactKeys,
            )
            const err = opts?.error
            const record: SpanRecord = {
              name,
              startTime,
              endTime,
              durationMs: endTime - startTime,
              app: config.app,
              env: config.env,
              context: merged,
              status: err ? 'error' : 'ok',
              ...(err
                ? {
                    error: {
                      name: err instanceof Error ? err.name : 'Error',
                      message: err instanceof Error ? err.message : String(err),
                    },
                  }
                : {}),
            }
            dispatch({ kind: 'span', record })
          },
        }
      },

      flush(): Promise<void> {
        return flushBuffer()
      },

      get sinks(): string[] {
        return [...sinkOrder]
      },

      addSink(sink: TelemetrySink): void {
        addSinkInternal(sink)
      },

      removeSink(name: string): void {
        if (sinks.has(name)) {
          sinks.delete(name)
          const idx = sinkOrder.indexOf(name)
          if (idx !== -1) sinkOrder.splice(idx, 1)
        }
      },
    }

    return logger
  }

  return makeLogger({})
}

export type { TelemetryPreset }

File: packages/flutter/lib/src/telemetry/telemetry_manager.dart
/// Telemetry manager — a 1:1 port of `@refraction-ui/logger`
/// `telemetry-manager.ts`. Manager/provider pattern, mirroring `createAI`.
library;

import 'dart:math';

import 'console_sink.dart';
import 'faro_engine.dart';
import 'lifecycle.dart';
import 'noop.dart';
import 'presets.dart';
import 'redact.dart';
import 'types.dart';

class _Entry {
  _Entry.log(this.logRecord) : kind = 'log', spanRecord = null;
  _Entry.span(this.spanRecord) : kind = 'span', logRecord = null;

  final String kind;
  final LogRecord? logRecord;
  final SpanRecord? spanRecord;
}

/// `createTelemetry` — creates a telemetry manager that fans records out to
/// registered sinks. Manager/provider pattern, mirroring `createAI`.
///
/// - `enabled: false` -> tree-shakeable noop (zero emissions, no engines).
/// - no `endpoint` -> console-only transport.
/// - `endpoint` set -> the (Wave 1) vendor engine is registered when a
///   binding/transport exists; console stays as a safe fallback until then.
///
/// The returned object IS a logger (root context = `{}`); `child()` derives
/// loggers with bound context (sessionId / interviewId / turnId / ...).
///
/// Identical API and structure across web/android/ios/desktop — platform
/// differences (app-exit flush wiring) are handled internally via a
/// conditional-import [LifecycleHook] behind one surface.
Telemetry createTelemetry(
  TelemetryConfig config, {

  /// Override the random source (for deterministic sampling in tests).
  double Function()? randomSource,

  /// Override the app-exit hook (for tests / custom hosts).
  LifecycleHook? lifecycleHook,
}) {
  if (config.enabled == false) {
    return createNoopTelemetry();
  }

  final preset = resolvePreset(config.env);
  final sampleRate = config.sampleRate ?? preset.sampleRate;
  final redactKeys = config.redactKeys ?? const <String>[];
  final rng = randomSource ?? Random().nextDouble;

  final sinks = <String, TelemetrySink>{};
  final sinkOrder = <String>[];
  final buffer = <_Entry>[];

  void addSinkInternal(TelemetrySink sink) {
    if (sinks.containsKey(sink.name)) {
      sinks[sink.name] = sink;
    } else {
      sinks[sink.name] = sink;
      sinkOrder.add(sink.name);
    }
  }

  // Console transport is always present as the zero-dependency baseline.
  addSinkInternal(createConsoleSink(ConsoleSinkOptions(pretty: preset.pretty)));

  // When an endpoint is configured, attempt to attach the vendor engine. The
  // construction is async + lazy so the optional vendor never becomes
  // required and vendor names stay out of the public API.
  if (config.endpoint != null) {
    final endpoint = config.endpoint!;
    // ignore: discarded_futures
    createFaroSink(FaroEngineOptions(app: config.app, endpoint: endpoint))
        .then((faro) {
          if (faro != null) addSinkInternal(faro);
        })
        .catchError((_) {
          // Vendor absent or init failed — console remains.
        });
  }

  bool shouldSample() {
    if (sampleRate >= 1) return true;
    if (sampleRate <= 0) return false;
    return rng() < sampleRate;
  }

  void deliver(_Entry entry) {
    for (final name in sinkOrder) {
      final sink = sinks[name];
      if (sink == null) continue;
      if (entry.kind == 'log') {
        sink.log(entry.logRecord!);
      } else {
        sink.span(entry.spanRecord!);
      }
    }
  }

  Future<void> flushBuffer() async {
    if (buffer.isNotEmpty) {
      final pending = List<_Entry>.from(buffer);
      buffer.clear();
      for (final entry in pending) {
        deliver(entry);
      }
    }
    await Future.wait(
      sinkOrder.map((name) => sinks[name]?.flush() ?? Future<void>.value()),
    );
  }

  void dispatch(_Entry entry) {
    if (preset.batch) {
      buffer.add(entry);
      if (buffer.length >= preset.batchSize) {
        // ignore: discarded_futures
        flushBuffer();
      }
      return;
    }
    deliver(entry);
  }

  // ---- app-exit beacon flush (production preset) --------------------------
  // Uniform across all targets: web maps to pagehide/visibilitychange,
  // non-web maps to process/lifecycle signals — handled internally behind
  // the single [LifecycleHook] surface.
  if (preset.beaconFlush) {
    final hook = lifecycleHook ?? LifecycleHook();
    hook.onExit(() {
      // ignore: discarded_futures
      flushBuffer();
    });
  }

  late final Telemetry Function(LogContext) makeLogger;

  makeLogger = (LogContext boundContext) {
    void emit(LogLevel level, String message, [LogContext? context]) {
      if (LEVEL_ORDER[level]! < LEVEL_ORDER[preset.minLevel]!) return;
      if (!shouldSample()) return;
      final merged = redact(<String, Object?>{
        ...boundContext,
        ...?context,
      }, redactKeys);
      final record = LogRecord(
        level: level,
        message: message,
        timestamp: DateTime.now().millisecondsSinceEpoch,
        app: config.app,
        env: config.env,
        context: merged,
      );
      dispatch(_Entry.log(record));
    }

    return _TelemetryImpl(
      emit: emit,
      makeChild: (ctx) =>
          makeLogger(<String, Object?>{...boundContext, ...ctx}),
      startSpanImpl: (name, attributes) {
        final startTime = DateTime.now().millisecondsSinceEpoch;
        var ended = false;
        return _SpanImpl(() => ended, (value) => ended = value, (opts) {
          final endTime = DateTime.now().millisecondsSinceEpoch;
          final merged = redact(<String, Object?>{
            ...boundContext,
            ...?attributes,
            ...?opts?.attributes,
          }, redactKeys);
          final err = opts?.error;
          final record = SpanRecord(
            name: name,
            startTime: startTime,
            endTime: endTime,
            durationMs: endTime - startTime,
            app: config.app,
            env: config.env,
            context: merged,
            status: err != null ? 'error' : 'ok',
            error: err != null
                ? SpanError(
                    name: (err is Error || err is Exception)
                        ? err.runtimeType.toString()
                        : 'Error',
                    message: err.toString(),
                  )
                : null,
          );
          dispatch(_Entry.span(record));
        });
      },
      flushImpl: flushBuffer,
      sinksImpl: () => List<String>.from(sinkOrder),
      addSinkImpl: addSinkInternal,
      removeSinkImpl: (name) {
        if (sinks.containsKey(name)) {
          sinks.remove(name);
          sinkOrder.remove(name);
        }
      },
    );
  };

  return makeLogger(<String, Object?>{});
}

class _SpanImpl implements Span {
  _SpanImpl(this._isEnded, this._setEnded, this._onEnd);

  final bool Function() _isEnded;
  final void Function(bool) _setEnded;
  final void Function(SpanEndOptions?) _onEnd;

  @override
  void end([SpanEndOptions? opts]) {
    if (_isEnded()) return;
    _setEnded(true);
    _onEnd(opts);
  }
}

class _TelemetryImpl implements Telemetry {
  _TelemetryImpl({
    required void Function(LogLevel, String, [LogContext?]) emit,
    required Telemetry Function(LogContext) makeChild,
    required Span Function(String, LogContext?) startSpanImpl,
    required Future<void> Function() flushImpl,
    required List<String> Function() sinksImpl,
    required void Function(TelemetrySink) addSinkImpl,
    required void Function(String) removeSinkImpl,
  }) : _emit = emit,
       _makeChild = makeChild,
       _startSpan = startSpanImpl,
       _flush = flushImpl,
       _sinks = sinksImpl,
       _addSink = addSinkImpl,
       _removeSink = removeSinkImpl;

  final void Function(LogLevel, String, [LogContext?]) _emit;
  final Telemetry Function(LogContext) _makeChild;
  final Span Function(String, LogContext?) _startSpan;
  final Future<void> Function() _flush;
  final List<String> Function() _sinks;
  final void Function(TelemetrySink) _addSink;
  final void Function(String) _removeSink;

  @override
  void debug(String message, [LogContext? context]) =>
      _emit(LogLevel.debug, message, context);

  @override
  void info(String message, [LogContext? context]) =>
      _emit(LogLevel.info, message, context);

  @override
  void warn(String message, [LogContext? context]) =>
      _emit(LogLevel.warn, message, context);

  @override
  void error(String message, [LogContext? context]) =>
      _emit(LogLevel.error, message, context);

  @override
  void fatal(String message, [LogContext? context]) =>
      _emit(LogLevel.fatal, message, context);

  @override
  Telemetry child(LogContext context) => _makeChild(context);

  @override
  Span startSpan(String name, [LogContext? attributes]) =>
      _startSpan(name, attributes);

  @override
  Future<void> flush() => _flush();

  @override
  List<String> get sinks => _sinks();

  @override
  void addSink(TelemetrySink sink) => _addSink(sink);

  @override
  void removeSink(String name) => _removeSink(name);
}
